### ISSUE TITLE
MQTT support - feature discussion

### ISSUE THREAD

**Issue by penfold42** [hyperion-project/hyperion.ng#483]
Would be a good thing. What functionality does it need ?
Set static colour
Set brightness
Set on/off
Start effect/clear effect

**Comment by WoLpH** [comment]
Most questions can probably be answered (and perhaps some code can be borrowed) from https://github.com/MrBramme/Mqtt-Hyperion-Remote

**Comment by brindosch** [comment]
We are talking about Hyperion as a MQTT Consumer (Subscriber) right?
This is a first candidate for "hyperion python plugins".
So this can be kept small as a python plugin that can be configured and interacts directly with Hyperion. Instead porting over to C++.

**Comment by WoLpH** [comment]
Yes, this would be a lot of work (and more error-prone) to implement in C++

**Comment by penfold42** [comment]
Yes, but configuration via webUI would be good too in the same way the python effects have a UI json file

**Comment by ebartz** [comment]
This feature would be really awesome. I am trying to get Hyperion to work with my setup which involves different protocols. I thought about a feature for a webhook but then searched for MQTT as many of my devices and workflows use this to communicate.

Basically what it would need for configuration is a IP/Hostname, Port, MQTT Topic, Number of Leds

It could then just send RGB Values tot the configured MQTT Topic. As message you could just simply send a json object containing the RGB values.

**Comment by Lord-Grey** [comment]
@ebartz Sorry, would you mind clarifying...
Are you talking about 
a) hyperion sends RGB values to an MQTT topic or
b) hyperion receives RGB values from an MQTT topic and displays them on a hyperion connected LED-Device?
Thank you!

**Comment by ebartz** [comment]
@Lord-Grey thanks for pointing me at this. My use-case would be option a. My lightning setup consists of some hue light controlled by deCONZ and a couple WS2812B stripes which are controlled by ESPhome. My whole setup is managed with home assistant. So what I need is a way to get the RGB values so that I can then use node red to actually control the individual lights.

But I also think, that option b is something that might be useful for others that want to control their hyperion setup and integrate it in their automations by simply using MQTT.

**Comment by accelle17** [comment]
+1 on this on controlling the lights via MQTT. This is supported mostly by all opensource automation system, (HASS, OpenHab, Tasmota / ESPHome lights).

**Comment by Dinth** [comment]
+1 but im especially interested in Hyperion sending RGB values to MQTT so i can use HA automation to set colour on (unsupported by Hyperion) RGB lights.

**Comment by m-seker** [comment]
> @ebartz Sorry, would you mind clarifying...
> Are you talking about
> a) hyperion sends RGB values to an MQTT topic or
> b) hyperion receives RGB values from an MQTT topic and displays them on a hyperion connected LED-Device?
> Thank you!

I'll implement case a

**Comment by m-seker** [comment]
Bad news, MQTT module of QT is under automation package 
https://doc.qt.io/QtMQTT/index.html

@Paulchen-Panther  @tpmodding @Lord-Grey  Other alternatives are https://github.com/eclipse/mosquitto and https://github.com/eclipse/paho.mqtt.c, both backed by Eclipse.

Mosquitto seems better maintained.
Any preference ?

**Comment by WoLpH** [comment]
Mosquitto is the broker (server), not the client so that's not an option in any case

Just my two cents on the subject, Hyperion already supports several other protocols which are more suited for live color changes.

I'm not saying it's not possible with MQTT but I would recommend another protocol simply because MQTT goes through a server first which means you have double the delay.

@ebartz instead of ESPHome I would suggest using WLED which supports both Home Assistant and Hyperion without issues. And beyond that, it's a far nicer interface for controlling leds. I love ESPHome and I use it a lot, but it's not great for controlling leds.

**Comment by WoLpH** [comment]
As for controlling hyperion from home assistant, there's already a library for that: https://www.home-assistant.io/integrations/hyperion/

**Comment by m-seker** [comment]
There is a library called libmosquitto which provides client functionality https://mosquitto.org/api/files/mosquitto-h.html

**Comment by WoLpH** [comment]
I stand corrected :)

**Comment by jes1417** [comment]
@WoLpH the library for HA is outdated and currently does not work with hyperion.ng.

**Comment by m-seker** [comment]
> @WoLpH the library for HA is outdated and currently does not work with hyperion.ng.

I just tested and seems to work fine ?

**Comment by jes1417** [comment]
For me the light toggle doesn't work in HA

https://github.com/home-assistant/core/issues/36471

**Comment by jes1417** [comment]
Also 

https://github.com/home-assistant/core/issues/35464

**Comment by jclsn** [comment]
Any news on this? I was hoping to be able to control my HA entities with hyperion

**Comment by renne** [comment]
What's the current state?
I want to control a Shelly RGBW2 to adjust the room lighting (RGBW strip) to the current image average color and brightness.

**Comment by Lord-Grey** [comment]
> What's the current state? I want to control a Shelly RGBW2 to adjust the room lighting (RGBW strip) to the current image average color and brightness.

MQTT is just a way of communication.
To have Shelly via MQTT you need to first have the Shelly protocol supported, plus MQTT capabilities.

Currently, this is not a priority, but you are welcome to implement a Shelly device.

**Comment by renne** [comment]
@Lord-Grey
Shelly devices already support MQTT. Hyperion just needs to register with the MQTT-broker and send [JSON-messages](https://shelly-api-docs.shelly.cloud/gen1/#shelly-rgbw2-color-mqtt) to set color/brightnes values.

**Comment by fredleger** [comment]
The way i see it is that you get to mqtt controled lights for other reasons than ambilight like but want to integreate hyperion using the same protocol.
Make perfect sense for my setup where i want to remove philips hue protocol.
